File: src/Web/Endpoints/BlogPosts.cs
using GanihuhStack.Application.BlogPosts.Commands.CreateBlogPost;
using GanihuhStack.Application.BlogPosts.Commands.DeleteBlogPost;
using GanihuhStack.Application.BlogPosts.Commands.UpdateBlogPost;
using GanihuhStack.Application.BlogPosts.Queries.GetBlogPostBySlug;
using GanihuhStack.Application.BlogPosts.Queries.GetBlogPosts;
using Microsoft.AspNetCore.Http.HttpResults;

namespace GanihuhStack.Web.Endpoints;

public class BlogPosts : IEndpointGroup
{
    public static string? RoutePrefix => "/api/blog-posts";

    public static void Map(RouteGroupBuilder groupBuilder)
    {
        groupBuilder.MapGet(GetBlogPosts);
        groupBuilder.MapGet(GetBlogPostBySlug, "{slug}");
        groupBuilder.MapPost(CreateBlogPost).RequireAuthorization();
        groupBuilder.MapPut(UpdateBlogPost, "{id}").RequireAuthorization();
        groupBuilder.MapDelete(DeleteBlogPost, "{id}").RequireAuthorization();
    }

    public static async Task<Ok<BlogPostDto>> GetBlogPostBySlug(string slug, ISender sender)
    {
        var result = await sender.Send(new GetBlogPostBySlugQuery(slug));
        return TypedResults.Ok(result);
    }

    public static async Task<Ok<List<BlogPostDto>>> GetBlogPosts(ISender sender)
    {
        var result = await sender.Send(new GetBlogPostsQuery());
        return TypedResults.Ok(result);
    }

    public static async Task<Created<AckDto>> CreateBlogPost(CreateBlogPostCommand command, ISender sender)
    {
        var id = await sender.Send(command);
        return TypedResults.Created($"/api/blogposts/{id}", new AckDto("Blog post created successfully.", id));
    }

    public static async Task<Ok<AckDto>> UpdateBlogPost(int id, UpdateBlogPostCommand command, ISender sender)
    {
        await sender.Send(command with { Id = id });
        return TypedResults.Ok(new AckDto("Blog post updated successfully.", id));
    }

    public static async Task<Ok<AckDto>> DeleteBlogPost(int id, ISender sender)
    {
        await sender.Send(new DeleteBlogPostCommand(id));
        return TypedResults.Ok(new AckDto("Blog post deleted successfully.", id));
    }
}

File: src/Web/ClientApp/src/services/blogPosts.ts
import apiFetch from './api'

export interface BlogPostDto {
  id: number
  title: string
  slug: string
  content: string
  excerpt: string | null
  tags: string | null
  isPublished: boolean
  publishedDate: string | null
  created: string
  lastModified: string
}

export interface AckDto {
  message: string
  id: number
}

export interface CreateBlogPostRequest {
  title: string
  content: string
  excerpt?: string
  tags?: string
}

export interface UpdateBlogPostRequest {
  title: string
  content: string
  excerpt?: string
  tags?: string
  slug?: string
  isPublished: boolean
}

export const getBlogPosts       = () => apiFetch<BlogPostDto[]>('/api/blog-posts')
export const getBlogPostBySlug  = (slug: string) => apiFetch<BlogPostDto>(`/api/blog-posts/${slug}`)
export const createBlogPost  = (data: CreateBlogPostRequest) => apiFetch<AckDto>('/api/blog-posts', { method: 'POST', body: JSON.stringify(data) })
export const updateBlogPost  = (id: number, data: UpdateBlogPostRequest) => apiFetch<AckDto>(`/api/blog-posts/${id}`, { method: 'PUT', body: JSON.stringify(data) })
export const deleteBlogPost  = (id: number) => apiFetch<AckDto>(`/api/blog-posts/${id}`, { method: 'DELETE' })
